### PR TITLE
Upgrade to AWS Rust SDK

### LINKED ISSUE #5895: aws sources: Support assuming roles directly in materialized
Users have an expectation that a correctly configured AWS `credentials` or `profile` file will cause systems to assume configured roles. Unfortunately Rusoto does not do this out of the box ([issue](https://github.com/rusoto/rusoto/issues/1758), [pr](https://github.com/rusoto/rusoto/pull/1842)) -- afaict it doesn't even provide a generic profile/credentials parser.

Adding an StsAssumedRoleProvider into our provider chain isn't a huge lift, but deciding whether we require users to specify a `role_arn` in the source constructor or do the right thing and extract the role_arn from a provided profile is important. If there are not any decent credentials parsers around then we will need to implement that as well. The above-linked PR looks reasonable to a first approximation, for this specific use case.

While S3 is in experimental mode we could do the hackier thing and require the role_arn inside the source's `WITH` clause, with an intention to never stabilize that option, which would unblock #5891 and similar use cases.

### LINKED ISSUE #8162: Switch from Rusoto to official AWS SDK
The official [AWS SDK for Rust](https://github.com/awslabs/aws-sdk-rust) is now at the point where we can consider switching to it over Rusoto. It has ["tier 1"](https://github.com/awslabs/aws-sdk-rust/blob/main/AWS_SERVICES_SUPPORTED.md) support for S3 and Kinesis, which are the only two AWS services that we use in this repository.

Note that while the official SDK comes with a big "don't use this in production" warning label, we can just ignore that. As the only remaining semi-active "maintainer" of Rusoto, I can tell you for sure that Rusoto has more bugs than the official SDK. As another data point, @alex-hunt-materialize recently ported the [k8s-eip-operator](https://github.com/MaterializeInc/k8s-eip-operator) that we use on cloud from Rusoto to the official SDK, and had a good experiencing doing so. Switching to the official SDK will also automatically fix #5895.

Filing this in the Sources & Sinks project since most of the work will need to take place there.

### LINKED ISSUE #8647: S3 sources: Support WebIdentity as a credentials provider
This can be done either by adding the [`WebIdentityProvider`](https://docs.rs/rusoto_sts/0.47.0/rusoto_sts/struct.WebIdentityProvider.html) to our credentials chain, or (preferably) by migrating to the new aws-sdk-rust.

This is necessary to support IAM role forwarding Kubernetes service accounts.

### PR DESCRIPTION
**Pull request by benesch**
Still very much noodling on this, but wanted to get up what I had. 

Fix #5895.
Fix #8162.
Fix #8647.

### CODE REVIEW

**Comment by alex-hunt-materialize** [comment]
This looks great! Looks like the tests are just failing because it's incomplete.

**Comment by benesch** [comment]
This is inching closer to completeness. All usage of Rusoto has been ported to the new AWS SDK, and everything compiles, but I suspect bugs abound.

**Comment by benesch** [comment]
Ok, I think I've shaken out the bugs here. In terms of review:

  * For the `persist` crate, @danhhz or @petrosagg could you take a look?
  * For the `dataflow` crate, @petrosagg or @quodlibetor could you take a look?
  * @alex-hunt-materialize if you want to double check this on behalf of cloud, please do! I think this should just do the right thing as long as users don't provide static credentials when they create an S3 or Kinesis source.
  * I'm not too stressed about getting a review on the `testdrive`, `s3-datagen`, or `perf-kinesis` crates.

**Review by danhhz — COMMENTED** [review]
src/persist LGTM

**Review by alex-hunt-materialize — APPROVED** [review]

File: src/aws-util/src/kinesis.rs [hunk 1]
```diff
@@ -62,11 +72,13 @@ pub async fn list_shards(
...
 pub async fn get_shard_ids(
-    client: &KinesisClient,
+    client: &Client,
     stream_name: &str,
 ) -> Result<impl Iterator<Item = String>, anyhow::Error> {
     let shards = list_shards(client, stream_name).await?;
-    Ok(shards.into_iter().map(|s| s.shard_id))
+    Ok(shards
+        .into_iter()
+        .map(|s| s.shard_id.unwrap_or_else(|| "".into())))
```
Comment (benesch) [inline src/aws-util/src/kinesis.rs:81]: From @petrosagg: this ought to return an error instead.

**Review by petrosagg — APPROVED** [review]

**Pull request merged**